### PR TITLE
[RFR] First commit of HA doc.

### PR DESCRIPTION
**Pull request by manasdk**
Document outlining information important to understand what deploying StackStorm in HA means.

Steps for a reference HA deployment.

### CODE REVIEW

File: docs/source/reference/ha.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+for information on how to partition sensors. At this point st2sensorcontainer's do not form a cluster and
+distribute work or take over new work if some nodes in the cluster disappear. It would however be possible to a
+sensor itself to be implemented with HA in mind so that the same sensor can be deployed on multiple nodes with
+the sensor managing active-active or active-passive. Providing some platform level HA support for sensors is
+likely to be an enhancement to StackStorm in future releases.
+
+
+st2rulesengine
+~~~~~~~~~~~~~~
+This is a dual purpose process - its main function is to evaluate rules when it sees TriggerInstances and
```
Comment (dzimine) [inline docs/source/reference/ha.rst]: I'm concerned on maintaining two sets of definitions of services, here and in the /install/overview. Most of the description here is color on how it operates in multi-node, but some give pretty good definitions of the services themselves. Can we keep the separation (definitions in overview, HA details/roles here)? If it's not easy, keep as is, but also put some good soundbites back to overview, too.

File: docs/source/reference/ha.rst [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+|st2| uses this to cache Actions, Rules and Sensor metadata which already live in the filesystem. All the content should
+ideally be source-control managed in preferrably a git repository. |st2| also stores operation data like ActionExecution,
+TriggerInstance etc. MongoDB supports `replica set high-availability <https://docs.mongodb.org/v2.4/core/replica-set-high-availability/>` which we recommend to provide a safe failover.
+
+Loss of connectivity to a MongoDB cluster will cause downtime for |st2|. However, once a replica MongoDB is brought back it
+should be quite possible to bring |st2| back to operational state by simply loading the content. Easy access to old ActionExecutions will be lost but all the data of old ActionExecution will still be available in audit logs.
+
+Postgresql
+~~~~~~~~~~
+Used primarily by ``mistral-api`` and ``mistral-server``. To deploy Postgresql in HA please see `documentation <http://www.postgresql.org/docs/9.4/static/high-availability.html>` provided by the Postgreql project.
```
Comment (dzimine) [inline docs/source/reference/ha.rst]: PostgreSQL

File: docs/source/reference/ha.rst [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+The data stored in Postgresql is operational for mistral therefore starting from a brand new Postgresql in case of loss
+of a cluster will bring automation services back instantly. Certainly there will be downtime while a new DB cluster is provisioned.
+
+RabbitMq
+~~~~~~~~
```
Comment (dzimine) [inline docs/source/reference/ha.rst]: RabbitMQ.

File: docs/source/reference/ha.rst [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+|st2| has been systematically built with High availability(HA) as a goal. The exact deployment
+steps to achieve HA depend on the specifics of the infrastructure in which |st2| is deployed. This
+guide covers a brief explanation on various |st2| services, how they interact and the external
+services necessary for |st2| to function.
+
+In the section :doc:`/install/overview` a detailed picture and explanation on how single-box deployments
+work is provided. Repeating the picture here to keep context and use as a reference to layer on some
+HA deployment specific details.
+
+.. figure :: /_static/images/st2-deployment-big-picture.png
```
Comment (dzimine) [inline docs/source/reference/ha.rst]: we'll need a bigger picture... eventually. I can help ... eventually.

File: docs/source/reference/ha.rst (hunk 1 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Agreed. Do not want to repeat for sure.

File: docs/source/reference/ha.rst (hunk 4 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: I went ahead and added a big picture. Its is not as nice and detailed as the picture from overview. Hopefully what I have created can work as a start that can be built-on and is yet acceptable in the short-term.

**Comment by dennybaa** [comment]
@manasdk awesome! Is the DFS story for storing packs going to be covered at this time?

**Comment by manasdk** [comment]
> Is the DFS story for storing packs going to be covered at this time?

No that falls into content management and likely we will keep it outside of StackStorm context

**Comment by dennybaa** [comment]
Yup, but it's worth mentioning that HA setup should bring up any kind of DFS to store packs and virtualenvs, otherwise task distribution won't really work since no tasks, no sensors, no workflows etc are distributed across cluster.

**Comment by manasdk** [comment]
> but it's worth mentioning that HA setup should bring up any kind of DFS to store packs and virtualenvs

Yes. It slipped my mind to add that piece. Thanks for the reminder I will make space for that in the docs

**Comment by Kami** [comment]
Nice architecture diagram :+1: 

It would be great if we can put a source (if you used some tool or online service) in this repo as well, otherwise it will just get out of sync once we add / change things.

File: docs/source/reference/ha.rst [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+.. source https://docs.google.com/drawings/d/1X6u8BB9bnWkW8C81ERBvjIKRfo9mDos4XEKeDv6YiF0/edit
+
+Components
+----------
+
+st2api
+^^^^^^
+This process host the REST API endpoints that serve requests from  WebUI, CLI and ChatOps. It maintains
+connections to MongoDB to store and retrieve information. It also connects to RabbitMQ to push various
+kind of messages onto the message bus. It is a gunicorn managed process which by default listens on
```
Comment (Kami) [inline docs/source/reference/ha.rst]: Maybe we could mention it's a Python WSGI app running under gunicorn, but that's not a big deal.

File: docs/source/reference/ha.rst [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+port ``9101`` and is front-ended by Nginx acting as a reverse proxy.
+
+n st2api processes can be behind a loadbalancer in an active-active configuration. Each of these
```
Comment (Kami) [inline docs/source/reference/ha.rst]: Extra leading `n`

**Comment by manasdk** [comment]
@Kami 

> It would be great if we can put a source (if you used some tool or online service) in this repo as well, otherwise it will just get out of sync once we add / change things.

https://github.com/StackStorm/st2docs/pull/88/files#diff-8652a753ac6bbec7b52abb55dea67ce5R197
points to the source. pretty much like the diagram in the overview section.

File: docs/source/reference/ha.rst [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+can be deployed on separate compute instances. If using the PAM authentication backend special care has to be
+taken to guarantee that all boxes on which an instance of st2auth runs should have the same users. Generally,
+all st2auth process should see the same identities, via some provider if applicable, for the system to work
+predictably.
+
+st2stream
+^^^^^^^^^
+This process exposes a server-sent event stream. Requires access to both MongoDB and RabbitMQ. It is also a
+gunicorn managed process, listens on port ``9102`` by default and is front-ended by Nginx acting as a reverse
+proxy. Clients like WebUI and ChatOps maintain a persistent connection with an st2stream process and receive
```
Comment (Kami) [inline docs/source/reference/ha.rst]: Yeah, maybe a good idea to also clarify on that and contrast it to st2api (short lived vs long lived / opened connections and requests).

That's something which needs to be taken into account when doing a custom deployments since you need to configure appropriate timeouts for lb, wsgi app server, etc.

File: docs/source/reference/ha.rst [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+update from the st2stream server.
+
+n st2stream process can be behind a loadbalance in an active-active configuration. Since clients maintain
```
Comment (Kami) [inline docs/source/reference/ha.rst]: loadbalance -> loadbalancer

File: docs/source/reference/ha.rst [hunk 9]
```diff
@@ -0,0 +1,301 @@
+:orphan:
+
+High availability deployment
+============================
+
+|st2| has been systematically built with High availability(HA) as a goal. The exact deployment
```
Comment (Kami) [inline docs/source/reference/ha.rst]: Maybe we can also some how call out that most components are horizontally scalable and adding more instances of those components increases the overall throughout (well, until RabbitMQ is saturated so right now this is our main bottle neck).

File: docs/source/reference/ha.rst [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+In a proper distributed setup it is recommended to setup Zookeeper or Redis to provide a distributed co-ordination
+layer. See :doc:`Policies </policies>`. Using a default file based co-ordination backend will not work as it would
+in a single box deployment.
+
+st2resultstracker
+^^^^^^^^^^^^^^^^^
+Tracks results of execution handed over to mistral. It requires access to MongoDB and RabbitMQ to perform its function.
+
+n st2resultstracker will co-operate with each other to perform work. At startup though there is a possibility
+of work however there will not be any negative consequences to this duplicate work. Specifically the jobs to track results
```
Comment (Kami) [inline docs/source/reference/ha.rst]: This part reads kinda weird, is there a word missing? possibility of duplicate work?

File: docs/source/reference/ha.rst [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+also get stored in the DB in case there is no worker to take over the work and this pattern makes all result trackers
```
Comment (Kami) [inline docs/source/reference/ha.rst]: Maybe we could say that a work around (although, as you have pointed out, it's not a big deal) is to start / restart those instances with some delay in between.

File: docs/source/reference/ha.rst [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+n st2api processes can be behind a loadbalancer in an active-active configuration. Each of these
+processes can be deployed on separate compute instances.
+
+st2auth
+^^^^^^^
+All authentication is managed by this process. This process needs connection to MongoDB and an authentication
+backend. See `Authentication backends <ref-auth-backends>` for more information. It is a gunicorn
+managed process which by default listens on port ``9102`` and is front-ended by Nginx acting as a reverse proxy.
+
+n st2auth processes can be behind a loadbalancer in an active-active configuration. Each of these processes
```
Comment (emedvedev) [inline docs/source/reference/ha.rst]: Leading `n` here as well.

File: docs/source/reference/ha.rst [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+with the |st2| API must be via the configured loadbalancer.
+
+Required dependencies
+---------------------
+Some HA recommendations for the dependencies required by |st2| components. Depending on the exact infrastructure these
+may not be suitable and would only serve as a suggestion.
+
+MongoDB
+^^^^^^^
+|st2| uses this to cache Actions, Rules and Sensor metadata which already live in the filesystem. All the content should
```
Comment (Kami) [inline docs/source/reference/ha.rst]: I would probably say something along the lines of that is mostly servers the purpose of storing meta data.

File: docs/source/reference/ha.rst (hunk 8 above)
Comment (emedvedev) [inline docs/source/reference/ha.rst]: And the `n` :stuck_out_tongue:

File: docs/source/reference/ha.rst [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+Following are the steps to provision a controller box on Ubuntu 14.04.
+
+Install required dependencies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Install ``MongoDB``, ``PostgreSQL`` and ``RabbitMQ``.
+
+  .. code-block:: bash
+
+        sudo apt-get install -y mongodb-server rabbitmq-server postgresql
```
Comment (Kami) [inline docs/source/reference/ha.rst]: I'm personally not a fan of having any of those "low level" instructions in this document since it's way too easy for those things to get out of sync with the reality.

File: docs/source/reference/ha.rst (hunk 6 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: n -> multiple. Since it is obviously confusing.

File: docs/source/reference/ha.rst (hunk 4 above)
Comment (dzimine) [inline docs/source/reference/ha.rst]: nice looking and precise picture. Great!

**Comment by dzimine** [comment]
I didn't proof-read, but overall LGTM.

File: docs/source/reference/ha.rst (hunk 13 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: It says exactly that - no?

File: docs/source/reference/ha.rst (hunk 1 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Fixed 2dbc87216335d4a0209e113a9037ba35ecd85981

File: docs/source/reference/ha.rst (hunk 9 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Fixed 1e93e3ffbbc7a072c21458c803d24af0e19ee6ad

File: docs/source/reference/ha.rst (hunk 10 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Fixed 1e93e3ffbbc7a072c21458c803d24af0e19ee6ad

File: docs/source/reference/ha.rst (hunk 11 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: fixed 1e93e3ffbbc7a072c21458c803d24af0e19ee6ad

File: docs/source/reference/ha.rst (hunk 12 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: fixed 1e93e3ffbbc7a072c21458c803d24af0e19ee6ad

File: docs/source/reference/ha.rst (hunk 14 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Yeah this is pretty much the same as the install docs in that case. It will get out of sync at some point and we will be on the hook to fix.

At this point my overwhelming intent is to provide an understanding about deploying StackStorm in HA and to provide concrete steps that can be referenced. To me things are concrete only when exact statements can be copy-pasted which is why it is my preference to have exact steps in the docs.

In the future I suspect we will remove all this from here and provide some pre-canned scripts, puppet modules or chef recipes that make things testable and easier to keep in sync. Until then this is something is better than nothing approach. 

TLDR; agreed this is not perfect, brittle but until we commit to provide some more robust mechanism this is it.

File: docs/source/reference/ha.rst [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+n st2auth processes can be behind a loadbalancer in an active-active configuration. Each of these processes
+can be deployed on separate compute instances. If using the PAM authentication backend special care has to be
+taken to guarantee that all boxes on which an instance of st2auth runs should have the same users. Generally,
+all st2auth process should see the same identities, via some provider if applicable, for the system to work
+predictably.
+
+st2stream
+^^^^^^^^^
+This process exposes a server-sent event stream. Requires access to both MongoDB and RabbitMQ. It is also a
+gunicorn managed process, listens on port ``9102`` by default and is front-ended by Nginx acting as a reverse
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: st2auth and st2stream use the same port by default? or copy pasta?

File: docs/source/reference/ha.rst [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+
+n st2resultstracker will co-operate with each other to perform work. At startup though there is a possibility
+of work however there will not be any negative consequences to this duplicate work. Specifically the jobs to track results
+also get stored in the DB in case there is no worker to take over the work and this pattern makes all result trackers
+pick up the same initial work set. Once this work set is exhausted all subsequents tasks are round-robined.
+
+st2notifier
+^^^^^^^^^^^
+This is a dual purpose process - its main function is to generate ``st2.core.actiontrigger`` and ``st2.core.notifytrigger``
+based on the completion of ActionExecution. The auxiliary purpose is to act as a backup scheduler for actions that may
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: I'm not sure this auxiliary purpose means.. backup scheduler?

File: docs/source/reference/ha.rst [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+        CREATE DATABASE mistral OWNER mistral;
+        EHD
+
+6. Add stable |st2| repos.
+
+  .. code-block:: bash
+
+        curl -s https://packagecloud.io/install/repositories/StackStorm/staging-stable/script.deb.sh | sudo bash
+
+7. Setup st2web and SSL termination. Follow :ref:`install webui and setup ssl<ref-install-webui-ssl-deb>`.
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: This part of the linked docs fails on the controller, since we didn't apt-get install st2.
```sudo cp /usr/share/doc/st2/conf/nginx/st2.conf /etc/nginx/sites-available/```

Could be helpful just to link to [the nginx config in the repo](https://github.com/StackStorm/st2/blob/master/conf/nginx/st2.conf).

Comment (manasdk) [inline docs/source/reference/ha.rst]: Hmm, kinda hard since the linked docs are for package based installs. Also, this reference install depends on the package based install to provide prescriptive as well as instructional steps. Our guess is that going forward packages will be the dominant installation method.

File: docs/source/reference/ha.rst (hunk 16 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Auxiliary to the processes main purpose of serving as notification. Its basically a nice way for us to say we abused this process and stuffed it with dual purposes!

File: docs/source/reference/ha.rst (hunk 15 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Oh ya copy pasta. Thanks!

File: docs/source/reference/ha.rst [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+8. Configuration for Nginx as loadbalancer for controller box can be found
+`here <https://gist.github.com/manasdk/fce14029900e533a385d#file-shared_st2_Nginx-conf>`__. With this configuration
+Nginx will loadbalance all requests between the two blueprint boxes ``st2-multi-node-1`` and ``st2-multi-node-2``.
+This includes requests to ``st2api``, ``st2auth`` and ``mistral-api``. Nginx also serves as the webserver for st2web.
+
+9. Install st2chatops following from :ref:`setup chatops<ref-setup-chatops-deb>`.
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: We don't have working st2api or st2cli at this point in the tutorial, which the directions for st2chatops assume.

File: docs/source/reference/ha.rst [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+  .. code-block:: bash
+
+        curl -s https://packagecloud.io/install/repositories/StackStorm/staging-stable/script.deb.sh | sudo bash
+
+2. Install all |st2| components and mistral.
+
+  .. code-block:: bash
+
+        `sudo apt-get install -y st2 st2mistral`
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: nite: extra backticks here that impede copy-pastability.

File: docs/source/reference/ha.rst [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+
+3. Update Mistral connection to PostgreSQL in `/etc/mistral/mistral.conf` by changing `database.connection` property.
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: Round-robin DNS would be the best way to set this up as HA (assuming HA postgres)?

File: docs/source/reference/ha.rst (hunk 19 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Fix already 0a5f33384b213b12b1b974ff96f6f512be0d53d7 :)

File: docs/source/reference/ha.rst (hunk 20 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Yes. Mistral only accepts 1 property for database connection so round-robin DNS would be needed.  I assume postgres will be active-active?

File: docs/source/reference/ha.rst (hunk 18 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Ugh. Yeah sorry about that again don't have a clean way to resolve this discrepancy. I guess skipping the first step of "Validate that chatops pack is installed, and a notification rule is enabled:" can be skipped from st2chatops instructions.

Rough edges!

File: docs/source/reference/ha.rst [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+
+4. Update Mistral connection to RabbitMQ in `/etc/mistral/mistral.conf` by changing `default.transport_url` property.
+
+4. Setup users as per :ref:`here<ref-config-ssh-sudo-deb>`. Make sure that user configuration on all boxes running ``st2auth`` is
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: I think you wanted this ref to point to [configure-authentication](https://docs.stackstorm.com/latest/install/deb.html#configure-authentication), not [ref-config-ssh-sudo-deb](https://docs.stackstorm.com/latest/install/deb.html#ref-config-ssh-sudo-deb)

File: docs/source/reference/ha.rst [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+   identical. This ensures consistent authentication from the entire |st2| install since the request to authenticate a user
+   can be forwarded by the loadbalancer to any of the ``st2auth`` processes.
+
+5. Use `shared st2 config <https://gist.github.com/manasdk/fce14029900e533a385d#file-st2-conf>`__ and replace `/etc/st2/st2.conf`.
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: need lots of round-robin DNS in this st2.conf too (assuming it works/reconnects with each of the pieces)

Comment (codyaray) [inline docs/source/reference/ha.rst]: Intentionally sending all syslog to the st2-multi-node-controller?

File: docs/source/reference/ha.rst (hunk 21 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Thanks!
https://github.com/StackStorm/st2docs/pull/104

File: docs/source/reference/ha.rst (hunk 22 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Yeah for centralized logging. The log strategy to use is totally up to you though - can leave logs on every box or even ship to logstash, loggly etc. I just picked one I liked for the purpose of this doc.

File: docs/source/reference/ha.rst (hunk 20 above)
Comment (Kami) [inline docs/source/reference/ha.rst]: And to make all the API service processes fully HA, user would also need to deploy multiple load-balancers / nginx instances and use round-robin DNS to load-balance between them :)

That should be preferred over sharing a single load-balancer / nginx instance.

File: docs/source/reference/ha.rst (hunk 16 above)
Comment (codyaray) [inline docs/source/reference/ha.rst]: Ha, sorry, I meant to say that I don't understand the backup scheduler itself. What's it used for?

File: docs/source/reference/ha.rst (hunk 20 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: Yes having multiple outer load balancers is ideal.  I assumed for sake of this document that the load balancing strategy will be user specific so did not talk much about possible approaches.

File: docs/source/reference/ha.rst (hunk 16 above)
Comment (manasdk) [inline docs/source/reference/ha.rst]: It is https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/notifier/scheduler.py

The scheduler recovers any delayed executions. It is a safety-valve in case execution do not get scheduled once they have been delayed due to some policy https://docs.stackstorm.com/policies.html

File: docs/source/reference/ha.rst (hunk 19 above)
Comment (codyaray) [inline docs/source/reference/ha.rst]: need to add nginx to this apt-gt install line too. (the blueprint boxes don't have it installed yet)

File: docs/source/reference/ha.rst [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Install ``MongoDB``, ``PostgreSQL`` and ``RabbitMQ``.
+
+  .. code-block:: bash
+
+        sudo apt-get install -y mongodb-server rabbitmq-server postgresql
+
+
+2. Fix listen address in `/etc/postgresql/9.3/main/postgresql.conf` and have PostgreSQL listen an interface that has an
```
Comment (codyaray) [inline docs/source/reference/ha.rst]: also need to do this for mongodb in `/etc/mongodb.conf` (binds to `127.0.0.1` by default, at least with 2.4.9 on ubuntu 14.04)
